### PR TITLE
Add calls to create and delete required Dynamo tables

### PR DESCRIPTION
**Pull request by brentnd**
Similar to [KCL](https://docs.aws.amazon.com/streams/latest/dev/kinesis-record-processor-ddb.html), kinsumer should provide the ability for the internal dynamo tables (`_clients`, `_checkpoints`, `_metadata`) to be created programmatically.

The new calls on `kinsumer` are `CreateRequiredTables()` and `DeleteTables()` which are opt-in so no existing functionality is changed.

No test functionality was changed but the new methods allowed a lot of existing test setup/cleanup code to be removed by using the new methods.

### CODE REVIEW

**Comment by brentnd** [comment]
Looking at the build failures. Ran `golint` but checking `gometalinter` now.

**Comment by brentnd** [comment]
I'm seeing these errors with and without my changes. Is it possible that `gometalinter` updates since the last build in May are now showing this as an issue?
```statsd/statsd.go:37::warning: Errors unhandled.,LOW,HIGH (gosec)
statsd/statsd.go:45::warning: Errors unhandled.,LOW,HIGH (gosec)
statsd/statsd.go:46::warning: Errors unhandled.,LOW,HIGH (gosec)
statsd/statsd.go:47::warning: Errors unhandled.,LOW,HIGH (gosec)
statsd/statsd.go:48::warning: Errors unhandled.,LOW,HIGH (gosec)
statsd/statsd.go:54::warning: Errors unhandled.,LOW,HIGH (gosec)
statsd/statsd.go:55::warning: Errors unhandled.,LOW,HIGH (gosec)
```
The linting does pass when using `gometalinter.v2`.

**Comment by brentnd** [comment]
https://github.com/alecthomas/gometalinter/pull/505 in gometalinter renamed `gas` to `gosec` so the `.travis.yml` file needed to be updated to disable it by the new name.

**Comment by garethlewin** [comment]
FWIW the reason we didn't want to auto create the tables were so that people could setup alarms and the like. Reading code now.

File: config.go [hunk 1]
```diff
@@ -81,6 +90,13 @@ func (c Config) WithStats(stats StatReceiver) Config {
 	return c
 }
 
+// WithDynamoCapacity returns a Config with a modified dynamo capacities
+func (c Config) WithDynamoCapacity(readCapacity, writeCapacity int64) Config {
```
Comment (garethlewin) [inline config.go]: Would you mind breaking this out into two separate with calls please?

File: kinsumer.go [hunk 2]
```diff
@@ -234,6 +234,42 @@ func (k *Kinsumer) dynamoTableActive(name string) error {
...
+// dynamoTableExists returns an true if the given table exists
+func (k *Kinsumer) dynamoTableExists(name string) bool {
+	_, err := k.dynamodb.DescribeTable(&dynamodb.DescribeTableInput{
+		TableName: aws.String(name),
+	})
+	return err == nil
+}
+
+// dynamoCreateTable creates a table with the given name and distKey
+func (k *Kinsumer) dynamoCreateTable(name, distKey string) error {
```
Comment (garethlewin) [inline kinsumer.go]: aws-sdk-go has `waiters` (https://github.com/aws/aws-sdk-go/blob/master/service/dynamodb/waiters.go) that are a reliable way to know if the create worked or failed. Thoughts on adding these waiters to `dynamoCreateTable` and `dynamoDeleteTable` Alternatively to `CreateRequiredTables` ?

File: kinsumer.go [hunk 3]
```diff
@@ -388,3 +424,51 @@ func (k *Kinsumer) Next() (data []byte, err error) {
 
 	return data, err
 }
+
+// CreateRequiredTables will create the required dynamodb tables
+// based on the applicationName
+func (k *Kinsumer) CreateRequiredTables() error {
```
Comment (garethlewin) [inline kinsumer.go:460]: Thoughts on running these in go routines with errgroup? That would build the tables faster.

**Review by garethlewin — CHANGES_REQUESTED** [review]
Create work, thanks for this. I have a few questions in the review.

**Comment by brentnd** [comment]
I'll take a look at the comments and incorporate the requested changes tomorrow.

File: kinsumer.go (hunk 3 above)
Comment (brentnd) [reply kinsumer.go:460]: Done. Especially important now given the `waiter` in each create table call.

File: config.go (hunk 1 above)
Comment (brentnd) [reply config.go]: Done.

File: kinsumer.go (hunk 2 above)
Comment (brentnd) [reply kinsumer.go]: Done.

**Comment by brentnd** [comment]
Missed the new dependency on sync/errgroup.

**Comment by brentnd** [comment]
All of the change requests have been addressed.

**Comment by garethlewin** [comment]
@brentnd did you forget to push your changes? I do not see any new commits.

**Comment by brentnd** [comment]
I amended the original commit and force pushed.

**Comment by brentnd** [comment]
@GarethLewin  Any updates on this?

**Comment by garethlewin** [comment]
Sorry was away all of last week.

**Review by garethlewin — APPROVED** [review]

**Comment by garethlewin** [comment]
Just as a general comment, please next time don't force push over your changes, makes PRs much harder to read.

**Comment by brentnd** [comment]
Thanks @GarethLewin. Makes sense about the force push.

**Pull request merged**